src/main/java/uk/gov/ch/developer/docs/controller/developer/SignInController.java

```diff
@@ -0,0 +1,225 @@
+        this.clientId = reader.getMandatoryString("OAUTH2_CLIENT_ID");
+        this.redirectUri = reader.getMandatoryString("OAUTH2_REDIRECT_URI");
+    }
+
+    @GetMapping
+    public void getSignIn(HttpServletRequest httpServletRequest,
+            HttpServletResponse httpServletResponse) throws IOException {
+
+        Session chSession = (Session) httpServletRequest
+                .getAttribute(SessionHandler.CHS_SESSION_REQUEST_ATT_KEY);
+
+        // Redirect for user authentication (no scope specified)
+        redirectForAuth(chSession, httpServletRequest, httpServletResponse, null, false);
+    }
+
+    /**
+     * Redirects to a URI for the user to authenticate themselves
+     * 
+     * @param session The user's session, retrieved from context
+     * @param request
+     * @param response
+     * @param companyNumber The company number a user is viewing
+     * @param force Boolean determining whether to force re-authentication
+     * @throws IOException
+     */
+    protected void redirectForAuth(Session session, HttpServletRequest request,
+            HttpServletResponse response, String companyNumber, boolean force) throws IOException {
+
+        // Find the original requested url
+        StringBuilder originalRequestUrl = new StringBuilder(request.getRequestURL().toString());
+        String queryString = request.getQueryString();
+        if (queryString != null) {
+            originalRequestUrl.append("?").append(queryString);
+        }
+
+        // Set the scope
+        String scope = null;
+        if (companyNumber != null) {
+            scope = OAUTH_COMPANY_SCOPE_PREFIX + companyNumber;
+        }
+
+        // Generate and store a nonce in the session
+        Session sessionToUpdate = session;
+
+        if (sessionToUpdate == null) {
+            sessionToUpdate = new SessionImpl();
+        }
+
+        String nonce = generateNonce();
+        sessionToUpdate.getData().put(SessionKeys.NONCE.getKey(), nonce);
+
+        // Build oauth uri and redirect
+        String authoriseUri;
+        if (force) {
+            authoriseUri = createAuthoriseURIWithForceAndHint(originalRequestUrl.toString(), scope,
+                    nonce, getEmailFromSession(session));
+        } else {
+            authoriseUri = createAuthoriseURI(originalRequestUrl.toString(), scope, nonce);
+        }
+
+        response.sendRedirect(authoriseUri);
+    }
+
+    /**
+     * Retrieves email from a session if present, else returns an empty string
+     * 
+     * @param session User session from which to retrieve the email address
+     * @return email
+     */
+    private String getEmailFromSession(Session session) {
+
+        String email = "";
+
+        SignInInfo signInInfo = new SignInInfo();
+        if (session != null) {
+            signInInfo = session.getSignInInfo();
+        }
+
+        UserProfile userProfile = signInInfo.getUserProfile();
```

Comment:
Can you successfully obtain the email address with this code?

Reply:
Not yet - part of future work